dxr/plugins/python/indexers.py

```diff
@@ -35,19 +42,115 @@
         'properties': {
             'py_type': NEEDLE,
             'py_function': NEEDLE,
+            'py_derived': NEEDLE,
+            'py_bases': NEEDLE,
         },
     },
 }
 
 
+class TreeToIndex(TreeToIndexBase):
+    def __init__(self, *args, **kwargs):
+        super(TreeToIndex, self).__init__(*args, **kwargs)
+
+        self.python_path = getattr(self.tree, 'plugin_python_path', self.tree.source_folder)
+        self.class_bases = {}
+        self.derived_classes = defaultdict(list)
+        self.names = {}
+
+    @property
+    def unignored_files(self):
+        return unignored(self.tree.source_folder, self.tree.ignore_paths,
+                         self.tree.ignore_filenames)
+
+    def post_build(self):
+        for path in self.unignored_files:
+            if is_interesting(path):
+                self.analyze(path)
+
+        # Once we've got all the base classes, store the opposite
+        # direction: derived classes!
+        for class_name, bases in self.class_bases.items():
+            for base_name in bases:
+                base_name = normalize_name(self.names, base_name)
+                self.derived_classes[base_name].append(class_name)
+
+    def analyze(self, path):
+        """Parse a Python file and analyze it for import and class data
+        that we need to find the inheritance tree for classes later on.
+
+        """
+        module_path = path_to_module(self.python_path, path)
+        syntax_tree = ast.parse(file_contents(path, 'utf-8'))
+        for node in ast.walk(syntax_tree):
+            # Save the base classes of any class we find.
+            if isinstance(node, ast.ClassDef):
+                class_path = module_path + '.' + node.name
+                bases = []
+                for base in node.bases:
+                    base_name = convert_node_to_name(base)
+                    if base_name:
+                        bases.append(module_path + '.' + base_name)
+                self.class_bases[class_path] = bases
```

Comment:
Later, if this ends up running out of RAM on large projects, you might find needles_from_graph() educational. There we compress a map much like this one by expressing it as a graph. That way, you don't store "C→B→A" reduntantly if you store both "D→C→B→A" and "E→C→B→A" (where "→" means "is a subclass of").